### PR TITLE
Fix payment request buttons for Chinese provinces in Chrome

### LINKED ISSUE #1065: Cannot place order with payment request buttons for Chinese provinces
**Describe the bug**
Cannot place an order with Stripe payment request buttons with a shipping address in China due to an error.

**To Reproduce**
1. Click on payment request button on your product page and type in a shipping address in China:

https://cld.wthms.co/wWGc5E 
Full Size: https://cld.wthms.co/wWGc5E

2. Receive an error:

https://cld.wthms.co/0xWmes 
Full Size: https://cld.wthms.co/0xWmes

`Shipping Province is not valid. Please enter one of the following: Yunnan / 云南, Beijing / 北京, Tianjin / 天津, Hebei / 河北, Shanxi / 山西, Inner Mongolia / 內蒙古, Liaoning / 辽宁, Jilin / 吉林, Heilongjiang / 黑龙江, Shanghai / 上海, Jiangsu / 江苏, Zhejiang / 浙江, Anhui / 安徽, Fujian / 福建, Jiangxi / 江西, Shandong / 山东, Henan / 河南, Hubei / 湖北, Hunan / 湖南, Guangdong / 广东, Guangxi Zhuang / 广西壮族, Hainan / 海南, Chongqing / 重庆, Sichuan / 四川, Guizhou / 贵州, Shaanxi / 陕西, Gansu / 甘肃, Qinghai / 青海, Ningxia Hui / 宁夏, Macau / 澳门, Tibet / 西藏, Xinjiang / 新疆
`

Related PHP Warning:

`PHP Warning:  preg_match(): Unknown modifier '&amp;' in /home/mik/woosites/woo-payment/wp-content/plugins/woocommerce-gateway-stripe/includes/payment-methods/class-wc-stripe-payment-request.php on line 987`

I tested it with Apple Pay too - the province needs to be typed manually there, but it still fails even if I use the exact same province names as in the error, such as `Yunnan`.

**Expected behavior**
To be able to place an order for any Chinese province with Stripe payment request button. 

**Additional context**
Raised in https://wordpress.org/support/topic/php-warnings-87/

Noting that I accidentally tested it with 4.1.16 as well - same error.

### PR DESCRIPTION
**Pull request by ricardo**
# Changes proposed in this Pull Request:

Fixes #1065 

**Some context on this PR:**

Chrome and Opera (both Chromium based browsers), offer a dropdown list of states/provinces when you add shipping or billing addresses for the Payment Request Buttons. - This is a problem, because WooCommerce expects the state to be formatted as an abbreviated state code.

For some countries like the United States or Brazil, Chrome returns an abbreviated state code like `CA` (California). But for other countries like Argentina or China, it returns a full state/province name.

I also tested Safari, Edge and Brave, which support the Payment Request API, and they don't provide a dropdown list of states. - You have to type it manually. So in this case, matching the full length state with the WC abbreviated state is required.

Currently, there is a function `normalize_state`, which loops through WC states and finds the state that matches the dropdown values in Chrome.

**Changes proposed:**

- Implement a fix for China in the state normalization function.
- Refactor state normalization when calculating shipping.

# Testing instructions

- Create a shipping zone in `Anhui / 安徽, China` or any other Chinese state, and add a flat rate of $1.
- Make sure Payment Request Buttons is enabled and you have the `* 4242` test card added to Chrome.
- Go to the HTTPS URL of a product page that has shipping enabled and click the `Buy now` button.
- Choose China as your shipping country and the state you have configured in your shipping zone (e.g Anhui), fill the Zipcode with a 6 digit number and the other address details with fake data.
- Notice the Flat rate $1 shipping method is applied.
- Click "Pay" and notice the payment goes through without any errors.
- Check the orders page and see if the shipping info is correct, specially the state.
- Check also if billing/shipping details are correct when you choose Chinese provinces or states in other countries.

-------------------
- [ ] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).

### CODE REVIEW

File: includes/payment-methods/class-wc-stripe-payment-request.php [hunk 1]
```diff
@@ -1056,30 +1053,70 @@ public function normalize_state() {
...
+		if ( $state && $country && is_array( $wc_valid_states ) && sizeof( $wc_valid_states ) > 0 ) {
+
+			// If it's already normalized, skip.
+			if ( in_array( $state, array_keys( $wc_valid_states ) ) ) {
+				return $state;
+			}
+
+			$match_from_state_input = false;
+
+			// China - Adapt dropdown values from Chrome and accept manually typed values like 云南.
```
Comment (diegocurbelo) [inline includes/payment-methods/class-wc-stripe-payment-request.php:1090]: Is it worth it to add a link to the valid states/regions for CN?

**Review by diegocurbelo — COMMENTED** [review]
I tested it in chrome and works as expected.

The code looks good to me 🚢 .

@ricardo, do we need to add a changelog and/or readme entry?

File: includes/payment-methods/class-wc-stripe-payment-request.php (hunk 1 above)
Comment (ricardo) [reply includes/payment-methods/class-wc-stripe-payment-request.php:1090]: https://github.com/woocommerce/woocommerce/blob/master/i18n/states.php

This link? I think it is worth it.

**Comment by ricardo** [comment]
Thanks for reviewing this! I'm going to add a changelog entry too 👍

**Comment by ricardo** [comment]
@diegocurbelo 

I added the link as a comment and a changelog entry.

**Review by diegocurbelo — APPROVED** [review]
LGTM 🚢
